lib/msal-node/src/client/ManagedIdentityApplication.ts

```diff
@@ -0,0 +1,270 @@
+    createClientAuthError,
+} from "@azure/msal-common";
+import { ServiceFabric } from "./ManagedIdentitySources/ServiceFabric";
+import { AppService } from "./ManagedIdentitySources/AppService";
+import { CloudShell } from "./ManagedIdentitySources/CloudShell";
+import { AzureArc } from "./ManagedIdentitySources/AzureArc";
+import { Imds } from "./ManagedIdentitySources/Imds";
+import {
+    ManagedIdentityConfiguration,
+    ManagedIdentityNodeConfiguration,
+    buildManagedIdentityConfiguration,
+} from "../config/Configuration";
+import { version, name } from "../packageMetadata.js";
+import { ManagedIdentityRequest } from "../request/ManagedIdentityRequest";
+import { ManagedIdentityId } from "../config/ManagedIdentityId";
+import { CryptoProvider } from "../crypto/CryptoProvider";
+
+/**
+ * Class to initialize a managed identity and identify the service
+ */
+export class ManagedIdentityApplication {
+    private managedIdentityId: ManagedIdentityId;
+
+    private config: ManagedIdentityNodeConfiguration;
+
+    private logger: Logger;
+    private cacheManager: CacheManager;
+    private networkClient: INetworkModule;
+    private cryptoProvider: CryptoProvider;
+
+    private identitySource:
+        | ServiceFabric
+        | AppService
+        | CloudShell
+        | AzureArc
+        | Imds;
+
+    private fakeAuthority: Authority;
+
+    constructor(
+        managedIdentityId: ManagedIdentityId,
+        configuration?: ManagedIdentityConfiguration
+    ) {
+        this.config = buildManagedIdentityConfiguration(configuration || {});
+        this.managedIdentityId = managedIdentityId;
+
+        this.logger = new Logger(
+            this.config.system.loggerOptions,
+            name,
+            version
+        );
+
+        this.cacheManager = new DefaultStorageClass(
+            this.managedIdentityId.id,
+            DEFAULT_CRYPTO_IMPLEMENTATION,
+            this.logger
+        );
+
+        this.networkClient = this.config.system.networkClient;
+
+        this.cryptoProvider = new CryptoProvider();
+
+        this.identitySource = this.selectManagedIdentitySource();
+
+        const fakeAuthorityOptions: AuthorityOptions = {
+            protocolMode: ProtocolMode.AAD,
+            knownAuthorities: [Constants.DEFAULT_AUTHORITY],
+            cloudDiscoveryMetadata: "",
+            authorityMetadata: "",
+        };
+        this.fakeAuthority = new Authority(
+            Constants.DEFAULT_AUTHORITY,
+            this.networkClient,
+            this.cacheManager,
+            fakeAuthorityOptions,
+            this.logger
+        );
+    }
+
+    /**
+     * Acquire an access token from the cache or the managed identity
+     * @param managedIdentityRequest
+     * @param managedIdentityId
+     * @returns the access token
+     */
+    public async acquireToken(
+        managedIdentityRequest: ManagedIdentityRequest,
+        managedIdentityId: ManagedIdentityId
+    ): Promise<AuthenticationResult | null> {
```

Comment:
It will return null if forceRefresh has been set to true, if there is no cached access token, or if the cached access token is expired. This is how the ClientCredentials flow functions as well.

Reply:
resolved, per our conversation